feat(ingress): aws alb

Hunter-Thompson opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

TODO:

- [x] Add tests

#### Summary
Add option for AWS ALB Ingress

See: 
1. https://kubernetes-sigs.github.io/aws-load-balancer-controller/v2.4/
2. https://kubernetes-sigs.github.io/aws-load-balancer-controller/v2.4/guide/ingress/annotations/

#### Issue Link

Fixes #293 

#### API

```yaml
apiVersion: installation.mattermost.com/v1beta1
kind: Mattermost
metadata:
  name: mm-example-full                         # Chose the desired name
spec:
  size: 5000users                               # Adjust to your requirements
  awsLoadBalancerController:
    enabled: true
    hosts: 
      - hostName: 'asd.example.com'
    certificateARN: 'asdasd'
    internetFacing: true
  version: 6.0.1
  licenseSecret: ""  
```

#### Release Note

```release-note
Adds an option to configure an AWS ALB instead of the default Nginx Ingress
```

mattermod commented:
Hello @Hunter-Thompson,

Thanks for your pull request! A Core Committer will review your pull request soon. For code contributions, you can learn more about the review process [here](https://developers.mattermost.com/contribute/getting-started/code-review/).

Per the [Mattermost Contribution Guide](https://developers.mattermost.com/contribute/getting-started/), we need to add you to the list of [approved contributors](https://docs.google.com/spreadsheets/d/1NTCeG-iL_VS9bFqtmHSfwETo5f-8MQ7oMDE5IUYJi_Y/pubhtml?gid=0&single=true) for the Mattermost project.

**Please help complete the Mattermost [contribution license agreement](http://www.mattermost.org/mattermost-contributor-agreement/)?**
Once you have signed the CLA, please comment with `/check-cla` and confirm that the CLA check is green.

This is a standard procedure for many open source projects.

Please let us know if you have any questions.

We are very happy to have you join our growing community! If you're not yet a member, please consider joining our [Contributors community channel](https://pre-release.mattermost.com/core/channels/tickets) to meet other contributors and discuss new opportunities with the core team.

Hunter-Thompson commented:
/check-cla

mattermod commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

gabrieljackson commented:
Hey @Hunter-Thompson, just checking in to see if you are still working on this PR. We have been keeping an eye on it while you have had it opened as a draft. Let us know if you want to chat about anything regarding this functionality. Thanks!

Hunter-Thompson commented:
@gabrieljackson went away for awhile. Il be working on it this week.

Quick question, has `apis/mattermost/v1alpha1` been deprecated in favor of `apis/mattermost/v1beta1`? 

Do I have to make changes in both places or is `v1beta1` fine?

gabrieljackson commented:
Hey @Hunter-Thompson, yes only updating `v1beta` is fine here. `v1alpha` is deprecated as you mentioned.

Let us know if you need any additional support with this.

Hunter-Thompson commented:
@gabrieljackson could you have a look at the API and let me know if youd like changes? If no changes are required - Il start writing unit tests.

Szymongib reviewed (commented):
Thanks for the PR!

Regarding the API, since it impacts both creation of service and ingress, I suppose it makes sense to keep it separate. I am a bit worried about complexity implications, but it does look like a useful feature.

gabrieljackson reviewed (commented):
@Hunter-Thompson let us know if you need any support making the change that @Szymongib proposed. Thanks!

Hunter-Thompson commented:
@gabrieljackson I've made the changes proposed. If the API looks good, iIcan start testing in my local Kubernetes cluster and also write unit tests.

Hunter-Thompson commented on pkg/mattermost/mattermost_v1beta.go:
```diff
@@ -142,6 +170,70 @@ func GenerateIngressV1Beta(mattermost *mmv1beta.Mattermost) *networkingv1.Ingres
+	if mattermost.Spec.AWSLoadBalancerController.CertificateARN != "" {
+		ingressAnnotations["alb.ingress.kubernetes.io/certificate-arn"] = mattermost.Spec.AWSLoadBalancerController.CertificateARN
+		ingressAnnotations["alb.ingress.kubernetes.io/ssl-redirect"] = "443"
+		ingressAnnotations["alb.ingress.kubernetes.io/listen-ports"] = `'[{"HTTP": 80}, {"HTTPS":443}]'`
```
This gets converted to 

```
    alb.ingress.kubernetes.io/listen-ports: '''[{"HTTP": 80}, {"HTTPS":443}]'''
```
and then Kubernetes complains about the extra quotes.. Is there a work around for this?

Hunter-Thompson replied:
fixed this in b479418

fmartingr commented on controllers/mattermost/mattermost/mattermost_test.go:
```diff
@@ -389,6 +391,132 @@ func TestCheckMattermost(t *testing.T) {
 	})
 }
 
+func TestCheckMattermostAWSIngress(t *testing.T) {
```
This general tests are good, but I would are more checks, for example:
- Enabling/Disabling the AWS Load Balancer
- Switching from AWS Load Balancer to an Ingress (and vice-versa), all resources regarding both should be taken care of

Hunter-Thompson replied:
Il work on better tests in the next commit.

fmartingr replied:
Hey @Hunter-Thompson , thanks for the fast response! Just to clarify: The first ones are good to have, the others are nice to have! Just wanted to be sure that we disable or remove any spec everything is properly taken care of.

fmartingr reviewed (commented):
Hey @Hunter-Thompson, thanks a lot for your contribution! It took me a while to get up to speed with everything so apologies for the delayed response! 

I've reviewed your contribution and it's good to go for the most part, but I've added some comments regarding naming conventions, some clauses I found confusing and testing coverage. Please check them out and reach out if you have any questions. 

Again, thank you for your time on this! Greatly appreciated.

Hunter-Thompson commented on controllers/mattermost/mattermost/mattermost.go:
```diff
@@ -209,13 +223,39 @@ func (r *MattermostReconciler) checkMattermostIngress(mattermost *mmv1beta.Matte
 	status *mmv1beta.MattermostStatus,
-	reqLogger logr.Logger) (reconcileStatus, error) {
-
+	reqLogger logr.Logger) (reconcileStatus, error,
```
yes, reverted

Hunter-Thompson commented on controllers/mattermost/mattermost/mattermost.go:
```diff
@@ -209,13 +223,39 @@ func (r *MattermostReconciler) checkMattermostIngress(mattermost *mmv1beta.Matte
+	var desired *networkingv1.IngressClass
+	desired = mattermostApp.GenerateALBIngressClassV1Beta(mattermost)
+
+	if mattermost.Spec.AWSLoadBalancerController == nil {
```
correct, ive removed the check first check since we are already checking if it is nil, and if it is nil, it should delete the ingress class.

Hunter-Thompson commented on pkg/mattermost/mattermost_v1beta.go:
```diff
@@ -142,6 +170,70 @@ func GenerateIngressV1Beta(mattermost *mmv1beta.Mattermost) *networkingv1.Ingres
+func GenerateALBIngressClassV1Beta(mattermost *mmv1beta.Mattermost) *networkingv1.IngressClass {
+	ingressClass := &networkingv1.IngressClass{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            "alb",
```
I took from the Spec on line 182, `ingress.k8s.aws/alb`. Im open to other suggestions though.

fmartingr replied:
I think the easiest way to move forward here is to use the name of the installation `mattermost.Name`, that way we know the ingress class belongs to a certain installation and we can treat that resource accordingly in the installations lifecycle. if we don't need it (because the spec now points to a custom one or because the ingress got disabled) we can just dispose of it but if we share the default ingress class between installations it would add complexity to the cleanup process, since we would need to check if there are other installations using it before trying its removal. Thoughts @gabrieljackson ?

Hunter-Thompson replied:
This is a good idea, Ive made the required changes.

Hunter-Thompson commented:
@fmartingr, thanks for review. Ive pushed a new commit addressing some of the points. Il start working on better tests soon.

fmartingr commented:
> @fmartingr, thanks for review. Ive pushed a new commit addressing some of the points. Il start working on better tests soon.

Awesome, thanks! Just re-request my review or ping me once you're done so I can take a look :)

Hunter-Thompson commented:
> > @fmartingr, thanks for review. Ive pushed a new commit addressing some of the points. Il start working on better tests soon.
> 
> Awesome, thanks! Just re-request my review or ping me once you're done so I can take a look :)

@fmartingr, I changed the tests a bit, and added a few new ones.

It covers:

1. disable aws lb and enable aws lb
2. check if resources are deleted properly after enabling and disabling

Could you review?

Hunter-Thompson commented:
Did you get a chance to review @gabrieljackson?

Hunter-Thompson commented:
bump

fmartingr commented:
Hey @mirshahriar, since Gabe is pretty busy, would you mind taking a look at this?

mirshahriar commented:
> Hey @mirshahriar, since Gabe is pretty busy, would you mind taking a look at this?

I'll have a look today.

mirshahriar commented on controllers/mattermost/mattermost/mattermost.go:
```diff
@@ -185,9 +190,14 @@ func (r *MattermostReconciler) checkMattermostRoleBinding(mattermost *mmv1beta.M
+	desired = mattermostApp.GenerateIngressV1Beta(mattermost)
 
-	if !mattermost.IngressEnabled() {
+	if mattermost.Spec.AWSLoadBalancerController != nil && mattermost.Spec.AWSLoadBalancerController.Enabled {
```
`AWSLoadBalancerEnabled` method will do the task

mirshahriar commented on controllers/mattermost/mattermost/mattermost.go:
```diff
@@ -209,13 +219,38 @@ func (r *MattermostReconciler) checkMattermostIngress(mattermost *mmv1beta.Matte
 
+func (r *MattermostReconciler) checkMattermostIngressClass(mattermost *mmv1beta.Mattermost, reqLogger logr.Logger) error {
+	var desired *networkingv1.IngressClass
+	desired = mattermostApp.GenerateALBIngressClassV1Beta(mattermost)
```
Do not need to declare `desired` separately. We can use `desired := `

mirshahriar commented on pkg/mattermost/mattermost_v1beta.go:
```diff
@@ -39,6 +39,12 @@ func GenerateServiceV1Beta(mattermost *mmv1beta.Mattermost) *corev1.Service {
 		"service.alpha.kubernetes.io/tolerate-unready-endpoints": "true",
 	}
 
+	if mattermost.Spec.AWSLoadBalancerController != nil && mattermost.Spec.AWSLoadBalancerController.Enabled {
```
`AWSLoadBalancerEnabled` method will do the task

mirshahriar commented on pkg/mattermost/mattermost_v1beta.go:
```diff
@@ -94,6 +100,26 @@ func configureMattermostService(service *corev1.Service) *corev1.Service {
 }
 
+func configureMattermostServiceNodePort(service *corev1.Service) *corev1.Service {
+	service.Spec.Ports = []corev1.ServicePort{
```
Instead of doing same thing as `configureMattermostService`, can we add a separate function to set port? 

Both `configureMattermostServiceNodePort` & `configureMattermostService` will call that function to set these ports.

mirshahriar commented on pkg/mattermost/mattermost_v1beta.go:
```diff
@@ -144,6 +170,70 @@ func GenerateIngressV1Beta(mattermost *mmv1beta.Mattermost) *networkingv1.Ingres
+		ingress.Spec.IngressClassName = pkgUtils.NewString(mattermost.Spec.AWSLoadBalancerController.IngressClassName)
+	}
+
+	if mattermost.Spec.AWSLoadBalancerController.IngressClassName == "" {
```
We can use `else` here.

mirshahriar commented on controllers/mattermost/mattermost/mattermost.go:
```diff
@@ -209,13 +219,38 @@ func (r *MattermostReconciler) checkMattermostIngress(mattermost *mmv1beta.Matte
+	var desired *networkingv1.IngressClass
+	desired = mattermostApp.GenerateALBIngressClassV1Beta(mattermost)
+
+	if mattermost.Spec.AWSLoadBalancerController == nil || mattermost.Spec.AWSLoadBalancerController.Enabled == false || mattermost.Spec.AWSLoadBalancerController.IngressClassName != "" {
```
I think if `.IngressClassName == ""`, then you want to delete.

Hunter-Thompson replied:
If `.IngressClassName != ""` I need to delete the ingress class created by the operator. Since the Ingress will use the IngressClassName provided by the spec.

mirshahriar commented on apis/mattermost/v1beta1/mattermost_utils.go:
```diff
@@ -113,6 +124,18 @@ func (mm *Mattermost) GetIngressHostNames() []string {
 	return hosts
 }
 
+func (mm *Mattermost) GetAWSLoadBalancerHostNames() []string {
```
I think we do not need this method. The caller can get the the list from spec.

Hunter-Thompson replied:
@mirshahriar I was following: https://github.com/mattermost/mattermost-operator/blob/63f7b1d59b18835be02b190aaeefba24972f90fc/pkg/mattermost/mattermost_v1beta.go#L136

mirshahriar commented on controllers/mattermost/mattermost/mattermost.go:
```diff
@@ -185,9 +190,14 @@ func (r *MattermostReconciler) checkMattermostRoleBinding(mattermost *mmv1beta.M
 
 func (r *MattermostReconciler) checkMattermostIngress(mattermost *mmv1beta.Mattermost, reqLogger logr.Logger) error {
-	desired := mattermostApp.GenerateIngressV1Beta(mattermost)
+	var desired *networkingv1.Ingress
```
why declared `desired` separately?

mirshahriar reviewed (commented):
I added a few comments.

mirshahriar reviewed (approved):
LGTM. Thank you for your time and work.
